### PR TITLE
Set up GH Action to verify stable source URLs

### PR DESCRIPTION
**Pull request by hobofan**
As suggested in https://github.com/bazelbuild/bazel-central-registry/issues/69#issuecomment-1090885327, this PR adds a tool and GH Action to verify that all the source URLs follow the format of stable URL as per https://github.com/bazel-contrib/SIG-rules-authors/issues/11#issuecomment-1029861300

There are two open questions for me:
- Should we also verify that the archive has a file ending of `.tar.gz`? There are currently some sources that utilize the `.zip` variant of archives.
- The lint fails right now, as I was unsure how to handle the preexisting unstable URLs. I started to manually migrate the URLs, but the first one I adjusted (of `rules_cc`), yielded an artifact that had a different checksum to the previous artifact, which would break the integrity hash. What's the best way to move forward here?

### CODE REVIEW

**Comment by meteorcloudy** [comment]
Thanks for helping us setting up this check, much appreciated!

File: tools/verify_stable_archives.py
Comment (Wyverald) [inline tools/verify_stable_archives.py:44]: nit: "an" unstable source url

File: tools/registry.py
Comment (Wyverald) [inline tools/registry.py:172]: nit: fix indent

**Review by Wyverald — APPROVED** [review]
thank you :)

**Comment by Wyverald** [comment]
> Should we also verify that the archive has a file ending of .tar.gz? There are currently some sources that utilize the .zip variant of archives.

I thought it's fine to have .zip files. Any reason you know of to discourage .zip?

> The lint fails right now, as I was unsure how to handle the preexisting unstable URLs. I started to manually migrate the URLs, but the first one I adjusted (of rules_cc), yielded an artifact that had a different checksum to the previous artifact, which would break the integrity hash. What's the best way to move forward here?

This is tricky. Since the registry is still in the alpha stage, we can make breaking changes and just update the integrity hash. If this were to happen after launch, we'd probably need a way to run the lint against newly changed module versions only, which would also address potential scalability concerns. I'm not too sure how easy it is to do that though.

File: tools/verify_stable_archives.py
Comment (hobofan) [reply tools/verify_stable_archives.py:27]: On that front, I wasn't sure how opinionated the central registry is. As we are looking to have stable URLs, are we also assuming that all source URLs are hosted on GitHub, or are we just trying to verify that of all the source URLs that are hosted on GitHub we are using the known stable form?

I adjusted it now to only check `https://github.com` URLs, and otherwise assume that the provided URL is stable.

**Comment by fmeum** [comment]
> > Any reason you know of to discourage .zip?
> 
> Not from my side, no. I was just wondering whether the extension being included in this comment ([#69 (comment)](https://github.com/bazelbuild/bazel-central-registry/issues/69#issue-1124082318)) had any particular relevance, especially since the reply from GitHub support didn't seem to mention the file extension as a requirement for getting a stable archive.

No, the particular archive format should be relevant for the stability guarantee. The original issue only affected zip files so `.tar.gz` seemed more reasonable at the time, but the answer by GitHub support seems to also cover `.zip`.

File: tools/verify_stable_archives.py
Comment (alexeagle) [reply tools/verify_stable_archives.py:44]: also please link to somewhere explaining the issue and what you should do to change your URL

Comment (hobofan) [reply tools/verify_stable_archives.py]: I added a bit more explanation and a link to the same comment as linked above. If there is another link that better explains both the problem and its resolution, I'd be happy to change it.

**Comment by hobofan** [comment]
I've adjusted all the URLs, where possible.

There are still two open offenses with `upb` and `boringssl`, for which there are no releases provided on GH, and with that no stable URLs exist. How do we want to go about those?

I've seen that @fmeum has previously employed the technique of creating a release on a fork, but I'm not sure who should provide the fork in these cases (both right now and going forward)? The maintainers of the module (which in these two cases are currently empty)?

File: modules/bazel_skylib/1.0.3/source.json
Comment (meteorcloudy) [inline modules/bazel_skylib/1.0.3/source.json:3]: @fmeum Do you know if url under `https://github.com/<org>/<repo>/releases/download` is deterministic or not?

Comment (meteorcloudy) [reply modules/bazel_skylib/1.0.3/source.json:3]: It feels strange that we cannot depend on them, also the content may actually be different that the tagged archive, for example `strip_prefix` isn't needed here.

Comment (fmeum) [reply modules/bazel_skylib/1.0.3/source.json:3]: > @fmeum Do you know if url under `https://github.com/<org>/<repo>/releases/download` is deterministic or not?

Yep, these were explicitly claimed to be stable.

Comment (meteorcloudy) [reply modules/bazel_skylib/1.0.3/source.json:3]: Thanks, then @hobofan there is no need to replace those URLs.

Comment (hobofan) [reply modules/bazel_skylib/1.0.3/source.json:3]: Given that everything under `/release/download` is user-uploaded, GH would have to actively have to temper with them for them not to be stable.

We could adjust the logic to have 3 states:
- `stable` (serves a consistent artifact, except for active interference by the maintainer like deleting the release or uploading a new artifact under the same name)
- `unstable` (may not serve the same artifact on consecutive requests; e.g. as it is dynamically produced in a non-deterministic way) - (bikeshedding: `volatile` might be a more appropriate naming 🙃 )
- `unknown` (anything else with unknown semantics)

We could then error for `unstable` URLs (as we can suggest a better `stable` alternative), and warn for `unknown` URLs.

Following that:
- Everything under `https://github.com/<org>/<repo>/releases/download` -> `stable`
- `https://github.com/<org>/<repo>/archive/refs/tags/<tag>.{tar.gz|zip}` -> `stable`
- Everything else under `https://github.com` -> `unstable`
- All other URLs: `unknown`

What do you think? I'd rather do it properly than too hastily, and than this could also serve as the basis for https://github.com/bazelbuild/bazel/issues/15128 in Bazel proper.

Comment (meteorcloudy) [reply modules/bazel_skylib/1.0.3/source.json:3]: Sounds good! 
However, we still have to think about how to deal with `https://github.com/<org>/<repo>/archive/$commit`. Maintaining a release on a fork sounds like too much extra work. My preference is also to give a warning on this, because I think in most case they are going to be stable, we also automatically mirror the archive to a bazel mirror. But of course, we should try to convince the original repo author to have releases or at least version tags, the current checked in version is only made up.

File: modules/rules_nodejs/4.5.1/source.json
Comment (alexeagle) [inline modules/rules_nodejs/4.5.1/source.json:4]: this is wrong, the original was a release artifact and shouldn't be changed. it's much smaller than the entire source repo

**Review by alexeagle — APPROVED** [review]

**Comment by hobofan** [comment]
I have reverted all the changes to the module versions where I changed the `/release/download` URLs, which according to the discussion (https://github.com/bazelbuild/bazel-central-registry/pull/84#discussion_r872147716) we can assume to be stable.

I've also adjusted the script to the 3 state logic of `stable`/`unstable`/`unknown`, where the script will error with a descriptive message for an `unstable` URL. I don't think there would be much use in warning for `unknown` URLs for a script that is mainly meant for CI purposes (and it would distract from the root cause of the CI test failing).

> There are still two open offenses with upb and boringssl, for which there are no releases provided on GH

This issue still persists.

> Maintaining a release on a fork sounds like too much extra work.

As much as I agree that we should make life as easy as possible for a module maintainer to encourage contributions, I don't think it's an unreasonable amount of effort to expect from a maintainer.

For automated release workflows for modules, tagging a release is a simple step to integrate (and already common for pre-module rulesets).
For manual module maintainers that are also the original project's maintainers, it's as simple as tagging a commit and pushing that tag (or doing it via the GH UI).
For manual module maintainers that main the module separate from the project, they additionally have to fork the project once and after that have to fetch upstream changes (which as well is straightforward from the GH UI).

All-in-all I think that's an okay amount of effort in exchange to get a more reproducible central registry.

**Comment by meteorcloudy** [comment]
> For manual module maintainers that main the module separate from the project, they additionally have to fork the project once and after that have to fetch upstream changes (which as well is straightforward from the GH UI).

I'm fine with this process, but the problem here is that we currently don't actually have dedicated module maintainers for upb and boringssl. I think we should file issues for those two to track the progress until 1) the original project maintainers start to have version tags, 2) or a module maintainer volunteers to do so in a fork.

In the check, we can special case them for now, otherwise I assume the check will fail and block all PRs after merging this?

**Comment by hobofan** [comment]
> In the check, we can special case them for now, otherwise I assume the check will fail and block all PRs after merging this?

Sounds good to me 👍 I've added special cases for those two, so merging isn't blocked anymore.

> I think we should file issues for those two to track the progress until 1) the original project maintainers start to have version tags, 2) or a module maintainer volunteers to do so in a fork.

I think for the scope of this, tracking this in #69.
That a good portion of the current modules are missing a maintainer right now is a separate topic, but I wasn't sure if I should open an issue for that, or if there was already something existing planned (by @bazel-contrib?) in terms of recruiting maintainers for those?

**Comment by meteorcloudy** [comment]
> Sounds good to me 👍 I've added special cases for those two, so merging isn't blocked anymore.

Thanks!

> I think for the scope of this, tracking this in https://github.com/bazelbuild/bazel-central-registry/issues/69 is fine.

You're right, this issue should be enough.

> That a good portion of the current modules are missing a maintainer right now is a separate topic, but I wasn't sure if I should open an issue for that, or if there was already something existing planned (by https://github.com/bazel-contrib?) in terms of recruiting maintainers for those?

Indeed, I think it's a good idea to open an issue to track this effort.

**Review by meteorcloudy — APPROVED** [review]
Thanks, again!

**Comment by hobofan** [comment]
Thanks for everyone involved for guiding me through this! :)

**Comment by meteorcloudy** [comment]
And it works like a charm: https://github.com/bazelbuild/bazel-central-registry/actions 😃

**Comment by fmeum** [comment]
> It seems the linter didn't run for some PRs: #96 Anyone know how to tweak the github action settings?

The action runs on `push`, which means only on a push to a branch in this repository - something contributors can't do pre-merge. I think that the action would need to trigger on `pull_request`.

**Comment by meteorcloudy** [comment]
I changed the setting to 
image
But I'm not seeing an `Approve and Run` button for any PRs.

Not sure how to change the action to trigger on `pull_request` 🤔

File: .github/workflows/verify_source_urls.yml
Comment (fmeum) [inline .github/workflows/verify_source_urls.yml:2]: @meteorcloudy I think that replacing `push` with `pull_request` here should make the action run when external contributors open or push to PR branches.

Comment (meteorcloudy) [reply .github/workflows/verify_source_urls.yml:2]: Oh, I see, thank you!

**Pull request merged**
